src/mame/fuuki/fuukifg2.cpp

```diff
@@ -358,39 +209,20 @@ void fuuki16_state::sound_command_w(u8 data)
 	machine().scheduler().perfect_quantum(attotime::from_usec(50)); // Fixes glitching in rasters
 }
 
-template<int Layer>
-void fuuki16_state::vram_w(offs_t offset, u16 data, u16 mem_mask)
-{
-	COMBINE_DATA(&m_vram[Layer][offset]);
-	m_tilemap[Layer]->mark_tile_dirty(offset / 2);
-}
-
-template<int Layer>
-void fuuki16_state::vram_buffered_w(offs_t offset, u16 data, u16 mem_mask)
-{
-	const int buffer = (m_vregs[0x1e / 2] & 0x40) >> 6;
-	COMBINE_DATA(&m_vram[Layer][offset]);
-	if ((Layer & 1) == buffer)
-		m_tilemap[2]->mark_tile_dirty(offset / 2);
-}
-
 void fuuki16_state::main_map(address_map &map)
 {
 	map(0x000000, 0x0fffff).rom();
 	map(0x400000, 0x40ffff).ram();
-	map(0x500000, 0x501fff).ram().w(FUNC(fuuki16_state::vram_w<0>)).share(m_vram[0]);
-	map(0x502000, 0x503fff).ram().w(FUNC(fuuki16_state::vram_w<1>)).share(m_vram[1]);
-	map(0x504000, 0x505fff).ram().w(FUNC(fuuki16_state::vram_buffered_w<2>)).share(m_vram[2]);
-	map(0x506000, 0x507fff).ram().w(FUNC(fuuki16_state::vram_buffered_w<3>)).share(m_vram[3]);
+	map(0x500000, 0x507fff).m(m_fuukitmap, FUNC(fuukitmap_device::vram_map));
 	map(0x600000, 0x601fff).mirror(0x008000).ram().share(m_spriteram);   // mirrored?
 	map(0x700000, 0x703fff).ram().w(m_palette, FUNC(palette_device::write16)).share("palette");
 	map(0x800000, 0x800001).portr("SYSTEM");
 	map(0x810000, 0x810001).portr("P1_P2");
 	map(0x880000, 0x880001).portr("DSW");
 	map(0x8a0001, 0x8a0001).w(FUNC(fuuki16_state::sound_command_w));
-	map(0x8c0000, 0x8c001f).ram().w(FUNC(fuuki16_state::vregs_w)).share(m_vregs);
-	map(0x8d0000, 0x8d0003).ram().share(m_unknown);
-	map(0x8e0000, 0x8e0001).ram().share(m_priority);
+	map(0x8c0000, 0x8c001f).rw(m_fuukitmap, FUNC(fuukitmap_device::vregs_r), FUNC(fuukitmap_device::vregs_w));
+	map(0x8d0000, 0x8d0003).rw(m_fuukitmap, FUNC(fuukitmap_device::unknown_r), FUNC(fuukitmap_device::unknown_w));
+	map(0x8e0000, 0x8e0001).rw(m_fuukitmap, FUNC(fuukitmap_device::priority_r), FUNC(fuukitmap_device::priority_w));
```

Comment:
Is the mapping similar enough in the two drivers that you could implement a device map function and install it like:
```c++
	map(0x8c0000, 0x8effff).m(m_fuukitmap, FUNC(fuukitmap_device::regs_map));
```
and avoid needing to know the details of the layout in each driver?